optimizations.qmd

```diff
@@ -1,90 +1,967 @@
 
-- Convolution Algorithms
-- MM Kernels
-- Layer Fusion
-- Node Elimination
-- Graph Rewriting
+This process typically involves analyzing the machine learning model to identify and remove weights, nodes, or layers that have little impact on the model's outputs. By selectively pruning a model in this way, the total number of parameters can be reduced significantly without substantial declines in model accuracy. The resulting compressed model requires less memory and computational resources to train and run while enabling faster inference times.
 
-## Model Compression {#sec-kd}
+Model pruning is especially useful when deploying machine learning models to devices with limited compute resources, such as mobile phones or TinyML systems. The technique facilitates the deployment of larger, more complex models on these devices by reducing their resource demands. Additionally, smaller models require less data to generalize well and are less prone to overfitting. By providing an efficient way to simplify models, model pruning has become a vital technique for optimizing neural networks in machine learning.
 
-Explanation: Model compression is crucial in reducing the computational complexity of deep learning models while preserving their predictive performance. This section delves into various techniques that facilitate the compression of models, making them lighter and more manageable for deployment on resource-constrained devices, thereby fostering quicker and more efficient AI implementations.
+There are several common pruning techniques used in machine learning, these include structured pruning, unstructured pruning, iterative pruning, bayesian pruning, and even random pruning. In addition to pruning the weights, one can also prune the activations. Activation pruning specifically targets neurons or filters that activate rarely or have overall low activation. There are numerous other methods, such as sensitivity and movement pruning. For a comprehensive list of methods, the reader is encouraged to read the following paper: ["A Survey on Deep Neural Network Pruning: Taxonomy, Comparison, Analysis, and Recommendations" (2023)]([https://arxiv.org/pdf/2308.06767.pdf](https://arxiv.org/pdf/2308.06767.pdf)).
 
-- Knowledge distillation
-- Tensor decomposition methods
-- Low-rank matrix factorization
-- Learned approximations of weight matrices
+So how does one choose the type of pruning methods? Many variations of pruning techniques exist where each varies the heuristic of what should be kept and pruned from the model as well the number of times pruning occurs. Traditionally, pruning happens after the model is fully trained, where the pruned model may experience mild accuracy loss. However, as we will discuss further, recent discoveries have found that pruning can be used during training (i.e., iteratively) to identify more efficient and accurate model representations.
```

Comment:
Following the previous comment I did, again, this section is quite clear. But, this part could be rephrased for better clarity and structure. It's not clear how one would go about choosing a pruning method from this description.

I would mention that there is trade-offs. In my personal experience, I fully understood the potential of pruning in assignment 2 (pruning reduced the size of the model over 50%, and the accuracy barely changed!). I would mention that, if you could include a reference or a image, that would be even better. 

A suggested re-phrased paragraph: 
"Choosing a pruning method often involves trade-offs between simplicity, effectiveness, and the nature of your specific application. [...]"